File: AndroidLab/ch5/src/main/java/com/example/ch5/section1/Test1_2Activity.java
package com.example.ch5.section1;

import android.os.Bundle;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import com.example.ch5.R;
import com.example.ch5.databinding.ActivityTest12Binding;

public class Test1_2Activity extends AppCompatActivity {

    ActivityTest12Binding binding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);

        binding = ActivityTest12Binding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        //animation 로딩..
        Animation anim = AnimationUtils.loadAnimation(this, R.anim.in);

        anim.setAnimationListener(new Animation.AnimationListener() {
            @Override
            public void onAnimationStart(Animation animation) {

            }

            @Override
            public void onAnimationEnd(Animation animation) {
                Animation anim = AnimationUtils.loadAnimation(Test1_2Activity.this, R.anim.move);
                //animation 종료후 마지막 프레임 상태를 유지할 것인지에 대한 설정
                anim.setFillAfter(true);
                binding.imageView3.startAnimation(anim);
            }

            @Override
            public void onAnimationRepeat(Animation animation) {

            }
        });


        binding.imageView3.startAnimation(anim);
    }
}

File: AndroidLab2/ch2/src/main/java/com/example/ch2/section1/Test1_2Activity.kt
package com.example.ch2.section1

import android.os.Bundle
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.PreferenceDataStoreFactory
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.preferencesDataStoreFile
import com.example.ch2.R

class Test1_2Activity : AppCompatActivity() {

    //코드에서 필요한 순간 직접 생성..
//    val dataStore: DataStore<Preferences> = PreferenceDataStoreFactory.create(
//        produceFile = { preferencesDataStoreFile("my_prefs")}
//    )

    //위처럼 객체를 생성할 수도 있지만 by 위임으로 생성하는 것이 일반적..
    //by 에 의해 생성하면 singleton 을 보장한다..


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_test12)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }
    }
}